File: toolkit/test/accessibility/Options.java
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This file incorporates work covered by the following license notice:
 *
 *   Licensed to the Apache Software Foundation (ASF) under one or more
 *   contributor license agreements. See the NOTICE file distributed
 *   with this work for additional information regarding copyright
 *   ownership. The ASF licenses this file to you under the Apache
 *   License, Version 2.0 (the "License"); you may not use this file
 *   except in compliance with the License. You may obtain a copy of
 *   the License at http://www.apache.org/licenses/LICENSE-2.0 .
 */

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.Properties;

/** Load from and save options into a file.
*/
class Options
    extends Properties
{
    public static Options Instance ()
    {
        if (saOptions == null)
            saOptions = new Options ();
        return saOptions;
    }

    public static void SetBoolean (String sName, boolean bValue)
    {
        Instance().setProperty (sName, Boolean.toString(bValue));
    }

    public static boolean GetBoolean (String sName)
    {
        return Boolean.getBoolean(Instance().getProperty (sName));
    }

    public static void SetInteger (String sName, int nValue)
    {
        Instance().setProperty (sName, Integer.toString(nValue));
    }

    public static int GetInteger (String sName, int nDefault)
    {
        String sValue = Instance().getProperty (sName);
        if (sValue == null)
            return nDefault;
        else
            return Integer.parseInt (sValue);
    }

    public void Load (String sBaseName)
    {
        FileInputStream fis = null;
        try
        {
            fis = new FileInputStream (ProvideFile(sBaseName));
            load (fis);
        }
        catch (IOException e)
        {
            // Ignore a non-existing options file.
        }
        finally
        {
            try
            {
                if (fis != null)
                    fis.close();
            }
            catch (IOException ex)
            {
            }
        }
    }

    public void Save (String sBaseName)
    {
        FileOutputStream fos = null;
        try
        {
            fos = new FileOutputStream (ProvideFile(sBaseName));
            store (fos, null);
        }
        catch (IOException e)
        {
        }
        finally
        {
            try
            {
                if (fos != null)
                    fos.close();
            }
            catch (IOException ex)
            {
            }
        }
    }

    private Options ()
    {
    }

    private File ProvideFile (String sBaseName)
    {
        return new File (
            System.getProperty ("user.home"),
            sBaseName);
    }

    private static Options saOptions = null;
}

File: vcl/source/printer/Options.cxx
/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This file incorporates work covered by the following license notice:
 *
 *   Licensed to the Apache Software Foundation (ASF) under one or more
 *   contributor license agreements. See the NOTICE file distributed
 *   with this work for additional information regarding copyright
 *   ownership. The ASF licenses this file to you under the Apache
 *   License, Version 2.0 (the "License"); you may not use this file
 *   except in compliance with the License. You may obtain a copy of
 *   the License at http://www.apache.org/licenses/LICENSE-2.0 .
 */

#include <comphelper/processfactory.hxx>

#include <vcl/printer/Options.hxx>

#include <com/sun/star/beans/PropertyValue.hpp>
#include <com/sun/star/beans/XPropertySet.hpp>
#include <com/sun/star/configuration/theDefaultProvider.hpp>
#include <com/sun/star/container/XNameAccess.hpp>
#include <com/sun/star/uno/XComponentContext.hpp>
#include <com/sun/star/lang/XMultiServiceFactory.hpp>

namespace vcl::printer
{
void Options::ReadFromConfig(bool i_bFile)
{
    bool bSuccess = false;
    // save old state in case something goes wrong
    Options aOldValues(*this);

    // get the configuration service
    css::uno::Reference<css::lang::XMultiServiceFactory> xConfigProvider;
    css::uno::Reference<css::container::XNameAccess> xConfigAccess;
    try
    {
        // get service provider
        css::uno::Reference<css::uno::XComponentContext> xContext(
            comphelper::getProcessComponentContext());
        // create configuration hierarchical access name
        try
        {
            xConfigProvider = css::configuration::theDefaultProvider::get(xContext);

            css::beans::PropertyValue aVal;
            aVal.Name = "nodepath";
            if (i_bFile)
                aVal.Value <<= u"/org.openoffice.Office.Common/Print/Option/File"_ustr;
            else
                aVal.Value <<= u"/org.openoffice.Office.Common/Print/Option/Printer"_ustr;
            xConfigAccess.set(xConfigProvider->createInstanceWithArguments(
                                  u"com.sun.star.configuration.ConfigurationAccess"_ustr,
                                  { css::uno::Any(aVal) }),
                              css::uno::UNO_QUERY);
            if (xConfigAccess.is())
            {
                css::uno::Reference<css::beans::XPropertySet> xSet(xConfigAccess,
                                                                   css::uno::UNO_QUERY);
                if (xSet.is())
                {
                    sal_Int32 nValue = 0;
                    bool bValue = false;
                    if (xSet->getPropertyValue(u"ReduceTransparency"_ustr) >>= bValue)
                        SetReduceTransparency(bValue);
                    if (xSet->getPropertyValue(u"ReducedTransparencyMode"_ustr) >>= nValue)
                        SetReducedTransparencyMode(static_cast<TransparencyMode>(nValue));
                    if (xSet->getPropertyValue(u"ReduceGradients"_ustr) >>= bValue)
                        SetReduceGradients(bValue);
                    if (xSet->getPropertyValue(u"ReducedGradientMode"_ustr) >>= nValue)
                        SetReducedGradientMode(static_cast<GradientMode>(nValue));
                    if (xSet->getPropertyValue(u"ReducedGradientStepCount"_ustr) >>= nValue)
                        SetReducedGradientStepCount(static_cast<sal_uInt16>(nValue));
                    if (xSet->getPropertyValue(u"ReduceBitmaps"_ustr) >>= bValue)
                        SetReduceBitmaps(bValue);
                    if (xSet->getPropertyValue(u"ReducedBitmapMode"_ustr) >>= nValue)
                        SetReducedBitmapMode(static_cast<BitmapMode>(nValue));
                    if (xSet->getPropertyValue(u"ReducedBitmapResolution"_ustr) >>= nValue)
                        SetReducedBitmapResolution(static_cast<sal_uInt16>(nValue));
                    if (xSet->getPropertyValue(u"ReducedBitmapIncludesTransparency"_ustr)
                        >>= bValue)
                        SetReducedBitmapIncludesTransparency(bValue);
                    if (xSet->getPropertyValue(u"ConvertToGreyscales"_ustr) >>= bValue)
                        SetConvertToGreyscales(bValue);

                    bSuccess = true;
                }
            }
        }
        catch (const css::uno::Exception&)
        {
        }
    }
    catch (const css::lang::WrappedTargetException&)
    {
    }

    if (!bSuccess)
        *this = aOldValues;
}
}

/* vim:set shiftwidth=4 softtabstop=4 expandtab: */
